### ISSUE TITLE
[UX] Add a cancel link in the views UI

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2516]
When I edit a view, then decide not to make any changes, I always scroll around for a while looking for a cancel link or button. But then realize there is not one, and click "Back" in my browser.  I think we could improve the user experience in Backdrop by adding this missing Cancel link.

**Comment by Dinornis** [comment]
When I edit a view in 1.6 I get a CANCEL button right next to the SAVE button at the top of the page.

**Comment by klonos** [comment]
@Dinornis the save and cancel buttons show up in the view edit form **only once you make a change**. What @jenlampton means is that we should have the cancel button available **no matter if there was something edited or not**.

PR upcoming...

**Comment by quicksketch** [comment]
@klonos With your PR applied (or in the sandbox site), making a change won't un-hide the submit button, so it makes it so that you cannot save the View at all! There's probably a line of JavaScript that also needs to be updated somewhere to show that button after a change is made.

**Comment by BWPanda** [comment]
Took a while of searching, but finally found the solution to the issue that @quicksketch mentioned. [PR](https://github.com/backdrop/backdrop/pull/2724) attached.

**Comment by klonos** [comment]
Looking good, but do you guys think that perhaps the button label should be something like "Back", instead of "Cancel"? "Cancel" without a "Save" paired to it feels kinda weird (unless in a dialog).

**Comment by jenlampton** [comment]
I like `Cancel` because it's consistent with the rest of the UI. I think another word might be confusing.

**Comment by klonos** [comment]
Fair enough. What about the redirect thing then?

**Comment by stpaultim** [comment]
@klonos or @jenlampton - I think the redirect thing makes sense, assuming it's not too difficult to do - and I'm assuming it's not. I might have someone to work on this, can anyone provide a quick pointer on how to add the "redirect" thing?

Unless @BWPanda has the time and inclination to update the current PR. 

I'm good with the PR as it is, but the redirect does make sense.

**Comment by BWPanda** [comment]
I tried doing the redirect, but couldn't work it out...

The 'Cancel' button is currently a form button (as opposed to a link). As such, when I try to get the `HTTP_REFERER` path to redirect to after pressing 'Cancel', it redirects back to the view form itself, since that's where you were when you clicked 'Cancel'.

This doesn't happen with other form (like nodes), I think because their 'Cancel' buttons are actually links, and they work differently. I tried making the views 'Cancel' button a link instead, but then you just give it a URL to go to, it can't run the function that clears any unsaved changes from the view.

So not sure this is going to be possible... Someone else is welcome to take a stab though.

**Comment by docwilmot** [comment]
@BWPanda forms like `node_form()` do this by preparing the redirect destination in the form, not after clicking the link/button. Would that not work here?

**Comment by quicksketch** [comment]
The current PR looks good as-is to solve the missing Cancel button. Maybe we make a follow-up? Besides the redirect location, I'm also inclined to add a status message along the lines of "Changes to the view %name have been discarded." So you get some kind of confirmation. Like @BWPanda says this is a button rather than a link, because Views have persistent storage while you're editing. Clicking the Cancel button actually does something in Views, not just leaving the page.

**Comment by BWPanda** [comment]
I like the idea of making the redirect a follow-up. I've updated the current PR to display an info message re. the changes being discarded ('info' because neither 'status' nor 'warning' seemed appropriate).

![image](https://user-images.githubusercontent.com/2385329/161645604-ad57e72e-8e5f-4d59-a135-eb2708f57867.png)

**Comment by stpaultim** [comment]
Works for me.

**Comment by indigoxela** [comment]
There's some problem with the Tugboat sandbox - it's expired, still seems to exist, but the password doesn't work anymore.

Also: the PR branch is quite behind, maybe a rebase also makes sense, anyway.

**Comment by jenlampton** [comment]
Changing to NW based on the latest comment

**Comment by stpaultim** [comment]
I rebased the PR from @BWPanda and created a new one with his work to see if we can move this forward. 

@quicksketch, @jenlampton, @indigoxela, @klonos, @Dinornis, @docwilmot

**Comment by argiepiano** [comment]
@stpaultim I tested your PR. I'm finding the same problem as [this comment from 2017](https://github.com/backdrop/backdrop-issues/issues/2516#issuecomment-286949814). When I make changes to an existing View, the Save button never appears in the View, and therefore it can't be saved.

**Comment by argiepiano** [comment]
Actually, after clearing caches, the Save button is now appearing. The modified JS script hadn't loaded after applying the patch, because of the js cache.

**Comment by argiepiano** [comment]
Thanks, @stpaultim. I left a comment in the PR.

**Comment by klonos** [comment]
We have moved to using links instead of buttons for "Cancel". So the cancel action we are introducing here should also be `'#type' => 'link'` instead of `'#type' => 'submit'`.

**Comment by klonos** [comment]
...I'm actually not sure about the value of the JS that shows/hides the submit elements in that form. We aren't doing that anywhere else AFAIk.

**Comment by argiepiano** [comment]
> We have moved to using links instead of buttons for "Cancel". So the cancel action we are introducing here should also be `'#type' => 'link'` instead of `'#type' => 'submit'`.

@klonos, this was already discussed above. We need to use a submit element because Views needs to do something before leaving the page. It gets a lot of info about the view via POST. A regular link only does a GET

See https://github.com/backdrop/backdrop-issues/issues/2516#issuecomment-1088013852

**Comment by klonos** [comment]
> @klonos, this was already discussed above. ...

Right. Sorry, I missed that. Thanks for pointing it out. We actually have the same thing going on in the "Manage blocks" form. I'm not sure if I want to file a separate issue to have those submit buttons styled as links for consistency or not 🤔 ...perhaps I'll let it go for now, and if it ever comes up in the future, then we can sort it out.

What do you guys think about getting rid of the JS that shows/hides the submit button? Less complexity and code to maintain, and it'll bring things in par with the "Manage blocks" form and other forms too.

**Comment by stpaultim** [comment]
@klonos I simply rebased the previous PR with this JS. I haven't really thought about or know what the alternative would be.

**Comment by klonos** [comment]
Right, so here's my thinking... the issue description says this:
>When I edit a view, then decide not to make any changes, I always scroll around for a while looking for a cancel link or button. But then realize there is not one, and click "Back" in my browser. I think we could improve the user experience in Backdrop by adding this missing Cancel link.

That ^^ is happening because JS is hiding the buttons. The UX bug with the current implementation of this JS is that it was hiding both the "Save" and the "Cancel" buttons (until there was any edit performed, in which case it would reveal the "dirt form" message + both buttons). Our solution to that problem is to adjust that JS to only be hiding the Save button, and to leave the "Cancel" button alone (always shown). My proposal is to simply rip all that JS hiding code out, and show both buttons all the time, like we are doing in the "Manage blocks" form. That would solve the initially reported problem + reduce code/complexity.

**Comment by kiamlaluno** [comment]
On other pages, the submission buttons are never hidden, even in the case there have not been changes in the shown data. I think it would make sense not to hide *Save* and *Cancel* on the view edit page too.

**Comment by argiepiano** [comment]
> My proposal is to simply rip all that JS hiding code out, and show both buttons all the time, like we are doing in the "Manage blocks" form. 

Makes sense to me.

**Comment by klonos** [comment]
@stpaultim it seems that we have some consensus here. Are you going to be able to work on this so we can knock it out of the way?(please see instructions in my previous comment - should be straight-forward).

**Comment by stpaultim** [comment]
@klonos I think I got it. Made changes to simply leave both buttons visible all the time. 

I don't have strong opinions about this, so if this blows up, I was just following orders. ;-)

Look forward to hearing from others. Mostly, did I do enough ripping out of code. I ripped out the related code that was visible in this PR. Is there anything I'm missing?

Sorry, Tugboat sandboxes don't seem to be working for me right now. Anyone have any ideas about that?

**Comment by stpaultim** [comment]
Sandbox is working again!

**Comment by yorkshire-pudding** [comment]
@stpaultim 
I've tested it and it works well.  One thing I'm not certain of; the message is the same whether you make changes or not:

> Changes to the view Promoted content have been discarded.

If the aim is to keep generic and not check whether or not we are actually discarding changes, then maybe something like:

> Any changes made to the Promoted content view have been discarded

Also think swapping the view name and word 'view' around reads slightly better

**Comment by stpaultim** [comment]
@yorkshire-pudding and @herbdool - Incorporated feedback regarding the cancel confirmation message. 

Also noticed that the old PR was hiding this message, by deleting line 28 in core/modules/views/js/ajax.js . By restoring that line into the PR, it seems to be working as expected.

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/0c1abaca-c84d-4a1f-96b9-cc666f4c402b)

This once again needs testing.

**Comment by quicksketch** [comment]
I re-tested the PR and everything worked great. Thanks everyone for your work on this issue! I merged https://github.com/backdrop/backdrop/pull/4599 into 1.x and 1.27.x.
